### PR TITLE
Cache git dependencies as wheels

### PR DESCRIPTION
**Pull request by maksbotan**
Currently, poetry install will clone, build and install every git dependency when it's not present in the environment. This is OK for developer's machines, but not OK for CI - there environment is always fresh, and installing git dependencies takes significant time on each CI run, especially if the dependency has C extensions that need to be built.

This commit builds a wheel for every git dependency that has precise reference hash in lock file and is not required to be in editable mode, stores that wheel in a cache dir and will install from it instead of cloning the repository again.

# Pull Request Check List

- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

This is a copy of #6896, which got automatically closed when `feature/wheel-installer-and-builder` branch was merged and deleted.

@neersighted @radoering may I ping you about this feature? 🙏 

If it's OK, I'll try to write a test.

### CODE REVIEW

**Comment by maksbotan** [comment]
@sdispater may I ask your attention about this? Since 1.4.0 was released, this change can be incorporated into master. IMO this is an important improvement.

File: src/poetry/installation/executor.py
Comment (radoering) [inline src/poetry/installation/executor.py:585]: Can we be sure that reference is always a hash or tag and never a branch?

File: src/poetry/installation/executor.py
Comment (radoering) [inline src/poetry/installation/executor.py:595]: We should extract this to method in `ArtifactCache` similar to `get_cached_archive_for_link`.

File: src/poetry/installation/executor.py
Comment (radoering) [inline src/poetry/installation/executor.py:622]: Can we move instead of copy? Alternatively, we could pass an optional destination to `_prepare_archive`.

File: src/poetry/installation/chef.py
Comment (radoering) [inline src/poetry/installation/chef.py:261]: `subdirectory` is relevant, too

**Review by radoering — CHANGES_REQUESTED** [review]
Please note #7621 and see my review comments.

Further, I wonder if it might make sense to mark the cache directories of git dependencies somehow for debugging purposes, e.g. by creating some `created_from_git_dependency` file? This would also give you an easy way to get rid of all cached wheels for git dependencies without nuking your complete artifacts cache. 🤔

File: src/poetry/installation/executor.py
Comment (maksbotan) [reply src/poetry/installation/executor.py:622]: I ended up using the second suggestions, refactoring artifacts cache a bit, making git cache use the same logic to find correct archive as is done for ordinary packages. This is more safe then just grabbing the first wheel found in the cache directory.

File: src/poetry/installation/chef.py
Comment (maksbotan) [reply src/poetry/installation/chef.py:261]: Thanks!

**Comment by maksbotan** [comment]
Thank you for the review, @radoering. I've addressed your comments in a separate commit, I'll squash everything before the merge.

> Further, I wonder if it might make sense to mark the cache directories of git dependencies somehow for debugging purposes, e.g. by creating some `created_from_git_dependency` file? This would also give you an easy way to get rid of all cached wheels for git dependencies without nuking your complete artifacts cache. 🤔

I think this would make sense. What is a proper way to touch a file in `poetry`, or should I just grab `shutil` or something for that?

**Review by maksbotan — COMMENTED** [review]
Some tests fail after my refactoring, I'll look at that later, if the overall approach is ok.

File: src/poetry/installation/executor.py
Comment (radoering) [reply src/poetry/installation/executor.py:607]: Sounds good to me.

**Comment by radoering** [comment]
> What is a proper way to touch a file in `poetry`, or should I just grab `shutil` or something for that?

[path.touch()](https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch)

**Comment by maksbotan** [comment]
@radoering i've addressed your comments and fixed failing tests.

File: src/poetry/installation/executor.py
Comment (radoering) [inline src/poetry/installation/executor.py:619]:
```suggestion
        if package.source_resolved_reference:
```

just in case it's an empty string

File: src/poetry/installation/executor.py
Comment (radoering) [inline src/poetry/installation/executor.py:641]: We can revert this change (and storing reference in a variable some lines above) since we only use `source_resolved_reference` for caching now.

File: src/poetry/installation/executor.py
Comment (radoering) [inline src/poetry/installation/executor.py:650]:
```suggestion
            package.source_resolved_reference
```

File: src/poetry/utils/cache.py
Comment (radoering) [inline src/poetry/utils/cache.py:250]: Since we add subdirectory only if it's present for links, I'd do the same here.

File: src/poetry/utils/cache.py
Comment (radoering) [inline src/poetry/utils/cache.py:322]: Not used anymore. Can be removed.

**Review by radoering — CHANGES_REQUESTED** [review]
LGTM after some minor changes and with some tests.

- in `test_cache.py`, we should add tests for `get_cache_directory_for_git` and `get_cached_archive_for_git`
- in `test_executor.py`, we probably should extend `test_executor_should_write_pep610_url_references_for_git` similar to `test_executor_should_write_pep610_url_references_for_wheel_urls` or `test_executor_should_write_pep610_url_references_for_non_wheel_urls`

**Review by maksbotan — COMMENTED** [review]
Hi @radoering!

I've addressed your comments and added some tests. I did not add a test for `get_cached_archive_for_git`, as it is a wrapper around `_get_cached_archive`, which is tested elsewhere anyway. Moreover, there was no test for `get_cached_archive_for_link` anyway.

**Comment by radoering** [comment]
> Moreover, there was no test for `get_cached_archive_for_link` anyway.

Actually, there are two tests for `get_cached_archive_for_link`. 😉 Apart from that, you're right. I just added a small smoke test that checks that none of the internal assertions is raised.

Further, I fixed the lost cleanup for editable git installs. Thereby, I noticed that the content of PEP 610 `direct_url.json` had been wrong for editable git installs and fixed that, too. Please take a look at my commits and verify that they do not break anything for you.

**Comment by maksbotan** [comment]
@radoering thank you! 

I've run a few tests with my production use case and this branch seems to work correctly.

Should I rebase these commits now? Or perhaps squash into one?

**Review by radoering — APPROVED** [review]

**Comment by maksbotan** [comment]
Thank you! Could you give some estimate when this can hit a release? I understand that it would be in 1.5.x series?

**Comment by radoering** [comment]
It will be in 1.5. I wouldn't expect it in April. 1.4.2 was released about a week ago and we didn't do too much for 1.5 until then.

**Comment by maksbotan** [comment]
Thanks.

**Pull request merged**
